sts-mock/src/token/tokenHandler.test.ts

```diff
@@ -1,41 +1,184 @@
+import { validateServiceTokenRequest } from "./validateServiceTokenRequest/validateServiceTokenRequest";
 
-describe("Token", () => {
-  describe("Given lambdaHandler is called", () => {
-    it("Returns 200 response", async () => {
-      const mockLoggingAdapter = new MockLoggingAdapter();
-      const dependencies = {
-        env: {},
-        logger: () => new Logger(mockLoggingAdapter, registeredLogs),
-      };
-      const result = await lambdaHandlerConstructor(
-        dependencies,
-        buildLambdaContext(),
-      );
-
-      expect(mockLoggingAdapter.getLogMessages()[0].logMessage.message).toBe(
-        "STARTED",
-      );
-      expect(mockLoggingAdapter.getLogMessages()[1].logMessage.message).toBe(
-        "COMPLETED",
-      );
-
-      expect(result).toStrictEqual({
-        headers: {
-          "Content-Type": "application/json",
-          "Cache-Control": "no-store",
-          Pragma: "no-cache",
-        },
-        statusCode: 200,
-        body: JSON.stringify({
-          access_token: "accessToken",
-          token_type: "Bearer",
-          expires_in: 3600,
-        }),
+describe("Token Handler", () => {
+  let mockLogger: MockLoggingAdapter<MessageName>;
+  let event: APIGatewayProxyEvent;
+  let dependencies: Dependencies;
+
+  const env = {
+    MOCK_STS_BASE_URL: "mockStsBaseUrl",
+    KEY_STORAGE_BUCKET_NAME: "mockKeyStorageBucketName",
+  };
+
+  beforeEach(() => {
+    event = buildTokenRequest(
+      "subject_token=testSub&scope=mock_service_name.mock_apiName.mock_accessLevel",
+    );
+
+    mockLogger = new MockLoggingAdapter();
+    dependencies = {
+      env,
+      logger: () => new Logger(mockLogger, registeredLogs),
+      validateServiceTokenRequest: validateServiceTokenRequest,
+      tokenSigner: () => new MockTokenSignerSuccessResult(),
+      keyRetriever: () => new MockKeyRetrieverSuccessResult(),
+    };
+  });
+
+  describe("Environment variable validation", () => {
+    describe.each(Object.keys(env))("Given %s is missing", (envVar: string) => {
+      it("Logs ENVIRONMENT_VARIABLE_MISSING and returns 500 Server Error", async () => {
+        dependencies.env = JSON.parse(JSON.stringify(env));
+        delete dependencies.env[envVar];
+
+        const result = await lambdaHandlerConstructor(
+          dependencies,
+          event,
+          buildLambdaContext(),
+        );
+
+        expect(mockLogger.getLogMessages()[1].logMessage.message).toStrictEqual(
+          "ENVIRONMENT_VARIABLE_MISSING",
+        );
+        expect(mockLogger.getLogMessages()[1].data).toStrictEqual({
+          errorMessage: `No ${envVar}`,
+        });
+        expect(result.statusCode).toStrictEqual(500);
+        expect(result.body).toStrictEqual(
+          '{"error":"server_error","error_description":"Server Error"}',
+        );
+      });
+    });
+  });
+
+  describe("Validate Service Token Request", () => {
+    describe("Given the request body is invalid (e.g. undefined)", () => {
```

Comment:
I think it would be good to look at covering the different permutations here please

e.g. Given there is no scope, Given there is no subject token etc.

Reply:
Yes, I had done that originally, but because I added those scenarios to validateServiceTokenRequest.test.ts, I didn't want to repeat them here, it didn't seem necessary because the logic in the handler is the same? What do you think?

Reply:
No need to duplicate - apologies!

Reply:
I'm wondering here: if we aren't mocking this validation functionality in the lambda - do you think the unit tests would be better suited to be in the lambda handler unit tests?

By moving these unit tests to the lambda unit test file, we can refactor the validateServiceTokenRequest.ts file as much as we want without changing a single unit test.

Reply:
I think that's a good idea. I was not 100% sure about where they should go. I will move them over now...

Reply:
Done